### PR TITLE
[REF] stock: add TestStockCommon inheritance to quant tests

### PR DESCRIPTION
**Pull request by pfertyk**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/tests/test_quant.py [hunk 1]
```diff
@@ -1,37 +1,30 @@
...
-            cls.env,
-            name='Pauline Poivraisselle',
-            login='pauline2',
-            email='p.p@example.com',
-            notification_type='inbox',
-            groups='stock.group_stock_user',
-        )
 
         cls.product = cls.env['product.product'].create({
             'name': 'Product A',
```
Comment (ticodoo) [inline addons/stock/tests/test_quant.py]: Maybe we can rename some of these products so they're not _all_ called 'Product A', especially since there's already a `cls.productA` with the name 'Product A' in `TestStockCommon`. Also, do we still need the basic `cls.product` now that `cls.productA` can be used?

File: addons/stock/tests/test_quant.py [hunk 2]
```diff
@@ -1,37 +1,30 @@
...
+
     @classmethod
     def setUpClass(cls):
-        super(StockQuant, cls).setUpClass()
+        super().setUpClass()
         cls.demo_user = mail_new_test_user(
             cls.env,
-            name='Pauline Poivraisselle',
-            login='pauline',
+            name='Pauline Demo Poivraisselle',
```
Comment (ticodoo) [inline addons/stock/tests/test_quant.py]: Nitpick (i.e. can be ignored): Maybe we can give her a new name so it doesn't overlap anymore with the `user_stock_user` one?

File: addons/stock/tests/test_quant.py [hunk 3]
```diff
@@ -1404,27 +1371,24 @@ def test_set_on_hand_quantity_tracked_product(self):
         }])
 
 
-class StockQuantRemovalStrategy(TransactionCase):
+class TestStockQuantRemovalStrategy(TestStockCommon):
+
     def setUp(self):
```
Comment (ticodoo) [inline addons/stock/tests/test_quant.py]: It's be better to use `setUpClass` instead of `setUp` unless the test won't run for some reason (e.g. cache issues and other rare edge cases). I forgot the specific reason(s) why, but at some point I learned it to be this way. If you do convert it (and it doesn't cause any issues), the `self.` should be replaced with `cls.`

File: addons/stock/tests/test_quant.py [hunk 4]
```diff
@@ -1404,27 +1371,24 @@ def test_set_on_hand_quantity_tracked_product(self):
...
         super().setUp()
-        self.least_package_strategy = self.env['product.removal'].search(
+        least_package_strategy = self.env['product.removal'].search(
             [('method', '=', 'least_packages')])
         self.product = self.env['product.product'].create({
             'name': 'Product',
             'is_storable': True,
             'categ_id': self.env.ref('product.product_category_goods').id,
```
Comment (ticodoo) [inline addons/stock/tests/test_quant.py]: It's possible we don't need this `categ_id` anymore (it used to be required until a couple saas's ago). If we don't, then we can reuse one of the common's product instead

**Review by ticodoo — APPROVED** [review]
Looks pretty good to me. Just a couple small comments that don't necessarily need to be applied if you don't feel like it for this round of test leaning

File: addons/stock/tests/test_quant.py (hunk 4 above)
Comment (pfertyk) [reply addons/stock/tests/test_quant.py]: This was my initial idea :sweat_smile: Unfortunately, removing this line causes 3 tests to fail :disappointed: I can investigate this further, if you want?

Comment (ticodoo) [reply addons/stock/tests/test_quant.py]: It is strange that they would fail. I think the product category shouldn't affect quants :thinking: might be worth investigating to see if something is misconfigured

File: addons/stock/tests/test_quant.py (hunk 3 above)
Comment (pfertyk) [reply addons/stock/tests/test_quant.py]: Converted to `setUpClass` without issues, thanks :+1:

File: addons/stock/tests/test_quant.py (hunk 1 above)
Comment (pfertyk) [reply addons/stock/tests/test_quant.py]: Replaced `self.product` with `self.productA` and improved the names of other products :+1:

File: addons/stock/tests/test_quant.py (hunk 2 above)
Comment (pfertyk) [reply addons/stock/tests/test_quant.py]: Name changed to Delphine Demoiselle ;)

File: addons/stock/tests/test_quant.py (hunk 4 above)
Comment (pfertyk) [reply addons/stock/tests/test_quant.py]: OK, I think I got it: without `categ_id` in the product, the setting of `removal_strategy_id` does not work (the default category is False). This causes different results in a few tests. I've added a comment explaining that `categ_id` is necessary here. Do you think this is sufficient?

File: addons/stock/tests/test_quant.py [hunk 5]
```diff
@@ -188,32 +163,32 @@ def test_get_available_quantity_9(self):
...
-        self.assertEqual(self.env['stock.quant']._get_available_quantity(self.product, self.stock_location), 2.0)
-        self.env['stock.quant']._update_available_quantity(self.product, self.stock_location, 1.0)
-        self.assertEqual(self.env['stock.quant']._get_available_quantity(self.product, self.stock_location), 3.0)
-        self.assertEqual(len(self.gather_relevant(self.product, self.stock_location)), 2)
+        self.assertEqual(self.env['stock.quant']._get_available_quantity(self.productA, self.stock_location), 2.0)
+        self.env['stock.quant']._update_available_quantity(self.productA, self.stock_location, 1.0)
+        self.assertEqual(self.env['stock.quant']._get_available_quantity(self.productA, self.stock_location), 3.0)
+        self.assertEqual(len(self.gather_relevant(self.productA, self.stock_location)), 2)
 
     def test_increase_available_quantity_3(self):
```
Comment (clesgow) [inline addons/stock/tests/test_quant.py:194]: I'm afraid this one won't run anymore though, as we're using a new warehouse within the tests, it won't find any existing quants and thus skip the test every single time.

But I don't know how we can solve that, as we need the records to be actually in the database if we want the other cursor being able to `SELECT` these :confused: 
Maybe in this *very specific case* we'd want to fetch actual real data, not the ones from our test warehouse.

Comment (ticodoo) [reply addons/stock/tests/test_quant.py:194]: This test and `test_decrease_available_quantity_3` already have issues/may be broken. I just tried running it in saas-18.4 with demo data and it always skips the test.

Also see: https://github.com/odoo/odoo/pull/218546#discussion_r2236339026 to see how this test appears to be totally broken in v16 already when it was reactivated (for that version only). It might be worthwhile checking if this test still makes sense to keep? Or if it needs to be rewritten to actually work.

Comment (pfertyk) [reply addons/stock/tests/test_quant.py:194]: Hi @clesgow and @ticodoo ! Sorry for the delay, I am working on this in between other tasks (which have proven to be more complex than expected 😅).

From what I gathered from other GitHub issues, our solutions are to either introduce 3 new DB cursors, or to extract a new module with test data always included. The latter doesn't really seem like a solution to me, so I would like to focus on the cursors approach.

There is a similar case in account: `TestSequenceMixinConcurrency` class with its 2 tests. I started rewriting our test to achieve something similar. I managed to use multiple cursors, but the test fails now because of a lack of update :thinking: I am trying to figure out what I did wrong, I will keep you posted.

Also, a note: the goal of these refactoring changes is to start using separate stock data in tests (at some point: switch to a dedicated company and stock user). However, this test checks for concurrency, so I think we can keep using the global data, as an exception. I tried to create all required models in this test alone, and with all our dependencies it was too difficult (e.g. a new warehouse tries to modify groups of the user ...). In most "normal" stock operations, I think we should test them as a stock manager or user, but if we're interested in concurrency only, I believe keeping the global data is completely justified.

Comment (pfertyk) [reply addons/stock/tests/test_quant.py:194]: On second thought @clesgow @ticodoo : could we extract these 2 tests into a separate class called `TestStockQuantConcurrency` , remove the `TestStockCommon` inheritance and keep using global data in them? That way, we could proceed with the refactoring, and solve this issue independently. I think we will end up using global data here anyway (and that's OK given the circumstances: were not testing normal stock operations, but rather the database itself). In that case, these concurrency test will be omitted in the refactoring process anyway.

What is your opinion?

Comment (pfertyk) [reply addons/stock/tests/test_quant.py:194]: After discussing this with @ticodoo we agreed to keep using the previous dataset in the concurrency tests, but without extracting them to a separate class. A `TODO` as added to indicate the problem, and @Aurelienvd was informed in a different PR.

@ticodoo @clesgow could you please take another look? Since your last review, at least one test was added and adjusted, so feel free to check the entire file :sweat_smile:
